_posts/2020-05-30-Metal3-Cluster-Deployment.md

```diff
@@ -0,0 +1,544 @@
+      Path:         /etc/yum.repos.d/kubernetes.repo
+      Permissions:  0644
+    Init Configuration:
+      Local API Endpoint:
+        Advertise Address:  
+        Bind Port:          0
+      Node Registration:
+        Kubelet Extra Args:
+          Node - Labels:  metal3.io/uuid={{ ds.meta_data.uuid }}
+        Name:             {{ ds.meta_data.name }}
+    Join Configuration:
+      Control Plane:
+        Local API Endpoint:
+          Advertise Address:  
+          Bind Port:          0
+      Discovery:
+      Node Registration:
+        Kubelet Extra Args:
+          Node - Labels:  metal3.io/uuid={{ ds.meta_data.uuid }}
+        Name:             {{ ds.meta_data.name }}
+    Post Kubeadm Commands:
+      mkdir -p /home/metal3/.kube
+      cp /etc/kubernetes/admin.conf /home/metal3/.kube/config
+      chown metal3:metal3 /home/metal3/.kube/config
+    Pre Kubeadm Commands:
+      ifup eth1
+      dnf update -y
+      dnf install -y ebtables socat conntrack-tools
+      dnf install python3 -y
+      dnf config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
+      setenforce 0
+      sed -i 's/^SELINUX=enforcing$/SELINUX=permissive/' /etc/selinux/config
+      dnf install docker-ce docker-ce-cli --disableexcludes=kubernetes --nobest -y
+      dnf install gcc kernel-headers kernel-devel keepalived device-mapper-persistent-data lvm2 -y
+      echo  "Installing kubernetes binaries"
+      curl -L --remote-name-all https://storage.googleapis.com/kubernetes-release/release/v1.18.0/bin/linux/amd64/{kubeadm,kubelet,kubectl}
+      chmod a+x kubeadm kubelet kubectl
+      mv kubeadm kubelet kubectl /usr/local/bin/
+      mkdir -p /etc/systemd/system/kubelet.service.d
+      curl -sSL "https://raw.githubusercontent.com/kubernetes/release/v0.2.7/cmd/kubepkg/templates/latest/deb/kubelet/lib/systemd/system/kubelet.service" | sed "s:/usr/bin:/usr/local/bin:g" > /etc/systemd/system/kubelet.service
+      curl -sSL "https://raw.githubusercontent.com/kubernetes/release/v0.2.7/cmd/kubepkg/templates/latest/deb/kubeadm/10-kubeadm.conf" | sed "s:/usr/bin:/usr/local/bin:g" > /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+      usermod -aG docker metal3
+      systemctl enable --now docker keepalived kubelet
+    Users:
+      Name:  metal3
+      Ssh Authorized Keys: ${SSH_KEY}
+      Sudo:  ALL=(ALL) NOPASSWD:ALL
+  Replicas:  1
+  Version:   v1.18.0
+Status:
+  Replicas:              1
+  Selector:              cluster.x-k8s.io/cluster-name=eko,cluster.x-k8s.io/control-plane=
+  Unavailable Replicas:  1
+  Updated Replicas:      1
+Events:                  <none>
+```
+
+```console
+kubectl get Metal3MachineTemplate ${CLUSTER_NAME}-controlplane -n metal3 
+kubectl get bmh -n metal3 -w   > You should see all the nodes that were created at the time of metal3 deployment, along with their current status as the provisioning progresses
+```
+
+> Once the provisioning is finished, let's get the host-ip : 
```

Comment:
Ok, but also mention (earlier) that the provisioning started...

I think this should explain the creation of the Machine / Metal3Machine, and the relationship with the BareMetalHost - and how the bare-metal-operator takes care of the provisioning.

Also it's probably worth mentioning that the provisioning does take some time, and add some suggestion on how to monitor it - proably through `virsh` on the host, but also through `kubectl get` commands on the management cluster to observe the various states that we cross.